zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/Schema.java

```diff
@@ -0,0 +1,366 @@
+  static final String TABLE_DEPENDENCIES = "dependencies";
+  static final String VIEW_TRACE_BY_SERVICE = "trace_by_service";
+
+  static final String DEFAULT_KEYSPACE = "zipkin3";
+  private static final String SCHEMA_RESOURCE = "/cassandra3-schema.cql";
+  private static final String INDEX_RESOURCE = "/cassandra3-index.cql";
+  private static final String INDEX_RESOURCE_3_0 = "/cassandra3-index-3.0.cql";
+
+  private Schema() {
+  }
+
+  static Metadata readMetadata(Session session) {
+    KeyspaceMetadata keyspaceMetadata = getKeyspaceMetadata(session);
+
+    Map<String, String> replication = keyspaceMetadata.getReplication();
+    if ("SimpleStrategy".equals(replication.get("class")) && "1".equals(
+        replication.get("replication_factor"))) {
+      LOG.warn("running with RF=1, this is not suitable for production. Optimal is 3+");
+    }
+
+    // In Cassandra 3.9+, you can create a SASI index on a materialized view
+    TableMetadata foo = keyspaceMetadata.getTable("trace_by_service_span");
+    boolean allowFilteringOnDuration = true;
+    if (foo != null) {
+      for (IndexMetadata index : foo.getIndexes()) {
+        if (index.getTarget().equals("duration")){
+          allowFilteringOnDuration = false;
+          break;
+        }
+      }
+    }
+    if (allowFilteringOnDuration) {
+      LOG.warn("Duration queries will allow filtering. Consider upgrading to Cassandra 3.8+");
+    }
+
+    String compactionClass = keyspaceMetadata.getTable("traces")
+        .getOptions().getCompaction().get("class");
+
+    return new Metadata(compactionClass, allowFilteringOnDuration);
+  }
+
+  static final class Metadata {
+    final String compactionClass;
+    final boolean allowFilteringOnDuration;
+
+    Metadata(String compactionClass, boolean allowFilteringOnDuration) {
+      this.compactionClass = compactionClass;
+      this.allowFilteringOnDuration = allowFilteringOnDuration;
+    }
+  }
+
+  static KeyspaceMetadata getKeyspaceMetadata(Session session) {
+    String keyspace = session.getLoggedKeyspace();
+    Cluster cluster = session.getCluster();
+    KeyspaceMetadata keyspaceMetadata = cluster.getMetadata().getKeyspace(keyspace);
+
+    if (keyspaceMetadata == null) {
+      throw new IllegalStateException(String.format(
+          "Cannot read keyspace metadata for give keyspace: %s and cluster: %s",
+          keyspace, cluster.getClusterName()));
+    }
+    return keyspaceMetadata;
+  }
+
+  static KeyspaceMetadata ensureExists(String keyspace, Session session) {
+    KeyspaceMetadata result = session.getCluster().getMetadata().getKeyspace(keyspace);
+    if (result == null || result.getTable("traces") == null) {
+      LOG.info("Installing schema {}", SCHEMA_RESOURCE);
+      applyCqlFile(keyspace, session, SCHEMA_RESOURCE);
+      try {
+        LOG.info("Installing index {}", INDEX_RESOURCE);
+        applyCqlFile(keyspace, session, INDEX_RESOURCE);
+      } catch (InvalidConfigurationInQueryException e) {
```

Comment:
here's the fallback to not use SASI on duration (when we can't due to version). Since indexes are in a separate file, we should be able to more easily maintain them than the manually created indexes of the old implementation.